File: 0581_shortest_unsorted_continuous_subarray/find_unsorted_subarray.py
from typing import List


def findUnsortedSubarray(nums: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    _min, _max, start, end, n = float("inf"), float("-inf"), -1, -1, len(nums)
    for i in range(n):
        _min = min(_min, nums[n - 1 - i])
        _max = max(_max, nums[i])
        if _max > nums[i]:
            end = i
        if _min < nums[n - 1 - i]:
            start = n - 1 - i
    if start == -1:
        return 0
    return end - start + 1

File: 0581_shortest_unsorted_continuous_subarray/find_unsorted_subarray.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public int findUnsortedSubarray(int[] nums) {
        int n = nums.length;
        int start = -1, end = -1;
        int max = Integer.MIN_VALUE, min = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            min = Math.min(min, nums[n - 1 - i]);
            max = Math.max(max, nums[i]);
            if (max > nums[i]) {
                end = i;
            }
            if (min < nums[n - 1 - i]) {
                start = n - 1 - i;               
            }
        }
        if (start == -1) { // array is already sorted
            return 0;
        }
        return end - start + 1;
    }
}